pkg/controller/identity/iamrolepolicyattachment/controller.go

```diff
@@ -125,90 +123,61 @@ func (e *external) Observe(ctx context.Context, mgd resource.Managed) (managed.E
+		}
+	}
+
 	cr.SetConditions(runtimev1alpha1.Available())
 
-	cr.UpdateExternalStatus(*attachedPolicyObject)
+	cr.Status.AtProvider = iam.GenerateRolePolicyObservation(*attachedPolicyObject)
 
 	return managed.ExternalObservation{
-		ResourceExists: true,
+		ResourceExists:   true,
+		ResourceUpToDate: cr.Spec.ForProvider.PolicyARN == cr.Status.AtProvider.AttachedPolicyARN,
 	}, nil
 }
 
 func (e *external) Create(ctx context.Context, mgd resource.Managed) (managed.ExternalCreation, error) {
-	cr, ok := mgd.(*v1alpha3.IAMRolePolicyAttachment)
+	cr, ok := mgd.(*v1beta1.IAMRolePolicyAttachment)
 	if !ok {
 		return managed.ExternalCreation{}, errors.New(errUnexpectedObject)
 	}
 
 	cr.SetConditions(runtimev1alpha1.Creating())
 
-	req := e.client.AttachRolePolicyRequest(&awsiam.AttachRolePolicyInput{
-		PolicyArn: aws.String(cr.Spec.PolicyARN),
-		RoleName:  aws.String(cr.Spec.RoleName),
-	})
+	req := e.client.AttachRolePolicyRequest(iam.GenerateAttachRolePolicyInput(&cr.Spec.ForProvider))
+	req.SetContext(ctx)
 
 	_, err := req.Send(ctx)
 
-	return managed.ExternalCreation{}, errors.Wrapf(err, errAttach, cr.Spec.PolicyARN, cr.Spec.RoleName)
+	return managed.ExternalCreation{}, errors.Wrapf(err, errAttach, cr.Spec.ForProvider.PolicyARN, cr.Spec.ForProvider.RoleName)
 }
 
 func (e *external) Update(ctx context.Context, mgd resource.Managed) (managed.ExternalUpdate, error) {
-	cr, ok := mgd.(*v1alpha3.IAMRolePolicyAttachment)
-	if !ok {
-		return managed.ExternalUpdate{}, errors.New(errUnexpectedObject)
-	}
-
-	// TODO(soorena776): add more sophisticated Update logic, once we
-	// categorize immutable vs mutable fields (see #727)
-
-	// there is not a dedicated update method, so a basic update is implemented below
-	// based on changes on PolicyArn:
-	// if the previously attached policy is different than what is stated in the spec,
-	// for update it needs to first attach the updated one, and then detach the previous one
-	if cr.Status.AttachedPolicyARN == "" || cr.Spec.PolicyARN == cr.Status.AttachedPolicyARN {
-		// update is only necessary if the PolicyArn in the Status is set and is different
-		// from the one in Spec
-		return managed.ExternalUpdate{}, nil
-	}
-
-	aReq := e.client.AttachRolePolicyRequest(&awsiam.AttachRolePolicyInput{
-		PolicyArn: aws.String(cr.Spec.PolicyARN),
-		RoleName:  aws.String(cr.Spec.RoleName),
-	})
-	if _, err := aReq.Send(ctx); err != nil {
-		return managed.ExternalUpdate{}, errors.Wrapf(err, errAttach, cr.Spec.PolicyARN, cr.Spec.RoleName)
-	}
-
-	dReq := e.client.DetachRolePolicyRequest(&awsiam.DetachRolePolicyInput{
-		PolicyArn: aws.String(cr.Status.AttachedPolicyARN),
-		RoleName:  aws.String(cr.Spec.RoleName),
-	})
-
-	if _, err := dReq.Send(ctx); err != nil {
-		return managed.ExternalUpdate{}, errors.Wrapf(err, errDetach, cr.Status.AttachedPolicyARN, cr.Spec.RoleName)
-	}
-
+	// PolivyARN is the only distinguishing field and on update to that, new policy is attached
```

Comment:
I think we need to let user know that their `spec` is out of sync with the external resource. However, returning an error here will cause immediate requeue, which I am not sure we want. @negz @muvaf any thoughts here?

Reply:
Is `PolicyARN` actually an immutable field? If so, I think we can skip doing anything in `Update`.

Letting the user know about the drift is challenging in AWS because update requests are constructed only with update-able parameters, hence API server wouldn't return an error like GCP `this field cannot be updated`. Ideally, validation should reject editing an immutable field.

Reply:
@hasheddan @muvaf In the current implementation, if the `PolicyARN` in `spec` is changed anyhow, a new policy is being created. And in the `Update` method, we have:
 `if cr.Status.AttachedPolicyARN == "" || cr.Spec.PolicyARN == cr.Status.AttachedPolicyARN`

I think we can check for the above condition in `Observe` itself, and decide either to create a new policy or update the existing.

Reply:
@sahil-lakhwani I am hesitant to delete and recreate an external object as part of the update logic of a reconciler because it feels like it violates that contract of the lifecycle of a custom resource being tied to the lifecycle of an external resource. @muvaf let me know if you think differently here. It sounds like the path forward for the time being is to just always return `true` for `ResourceUpToDate` if the `IAMRolePolicyAttachment` exists at all (per @muvaf [comment](https://github.com/crossplane/provider-aws/pull/141/files/acb886579b36a352e22e7201e6f3b4f98a8cb786#r390895464)), and keep the current functionality you have here of doing nothing in the `Update` method.

@muvaf I wonder if we should expose something in `crossplane-runtime` that allows consumers to report back that "this resource is not up to date, but we are not trying to fix it". We could do this currently by manually updating status condition here, but I think `crossplane-runtime` would overwrite it after we returned an empty `managed.ExternalUpdate`.

Reply:
Yeah we don't delete anything unless actual deletion request comes in. @sahil-lakhwani this field should be marked as immutable and when we get around implementation of immutability, it will work. But for the time being, that's all we can do.

About the drift in general; there are three kinds of drifts:
* In progress: you changed something and the update process is going on but the property is still the same in external resource, like you increased the size of DB from 20GB to 400GB and `spec` says 400GB immediately while the process is going on. Not much to do here unless provider reports an unavailable state, then we could update the condition. But other than that, status is expected to show that an update is going on if possible.
* Immutable field changed by user. Specifically in AWS, `Modify` requests do not include those fields so we can't have provider return error in those cases. The best we could do is to prevent those fields from being edited.
* Update-able field changed but resulted in error. Provider tells you the error and we show that.

@hasheddan For the second case, we could investigate a mechanism where in `Observe` we compare `spec` with what exists in the provider and update a `condition` saying whether there is a drift without going into details like whether this was because an immutable field or an update is due etc. It's a bit challenging to implement this in a good way though (comparison is hard). Maybe use IsUpToDate check to set a condition? But that'll require full scan of fields opposed to selective scanning that current `IsUpToDate` functions do.